### ISSUE TITLE
COCO Evaluation performance mismatch

### ISSUE THREAD

**Issue by Dwrety** [IDEA-Research/GroundingDINO#96]
Hi. 

I have written my own COCO evaluation code. And I have tested the SwinT checkpoint against coco val2017 and I got the AP of 40.8, which is 8.0 less than the reported results. I am wondering if you can share the evaluation code? Or briefly describe the inference process and parameters? Much appreciated. 

```Evaluate annotation type *bbox*
DONE (t=30.64s).
Accumulating evaluation results...
DONE (t=8.52s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.408
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.593
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.437
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.232
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.444
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.588
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.335
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.557
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.612
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.400
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.659
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.804
```

What I did was I concat all 80 coco categories with a separation token of ". " (and "." also tested, no difference). 
The encoded query tokens has a length of 195 and padded to 256. Everything in the forward method is not changed. 
In the final prediction. The `pred_logits` has a shape of `[900, 256]`, I rank all prediction by it's max value on `dim=1` and picked the top 100 results `[100, 256]`.
Then I looked at each prediction, map the 256 scores back to 80 coco categories and take the maximum score. (Use the same method in GLIP).

Please point out if I am doing this wrong. Thank you.

**Comment by serser** [comment]
> The inference `transforms` follows the same concept in the release demo file. So shorter side 800 while max side 1333 and normalized, no padding.

Is it resolved finally?

**Comment by sachidanandvs** [comment]
Hi, @Dwrety were you able to get the reported results? I am also having trouble reproducing the results.

**Comment by Dwrety** [comment]
> > The inference `transforms` follows the same concept in the release demo file. So shorter side 800 while max side 1333 and normalized, no padding.
> 
> Is it resolved finally?

not yet. still trying tho

**Comment by serser** [comment]
@SlongLiu Hi, could you help us with the evaluation to reproduce the report results?

**Comment by HuixinSun** [comment]
Hi, @Dwrety is if possible for you to share the 40.8 mAP evaluation code? I am also trying to write the evaluation myself but got a much lower mAP.

**Comment by SlongLiu** [comment]
Hi @HuixinSun @serser @Dwrety . I provide a script to reproduce the 48.5 AP on COCO zero-shot.  See https://github.com/IDEA-Research/GroundingDINO#coco-zero-shot-evaluations for more details. Hope the file is helpful.

**Comment by charliess123** [comment]
> Hi @HuixinSun @serser @Dwrety . I provide a script to reproduce the 48.5 AP on COCO zero-shot. See https://github.com/IDEA-Research/GroundingDINO#coco-zero-shot-evaluations for more details. Hope the file is helpful.

Hi, I have tried your script and I find out that there is a category id mapping in the code, which make the whole category number increasing to 91. But coco only got 80 categories. What's that for?

**Comment by CConory** [comment]
> Hi @HuixinSun @serser @Dwrety . I provide a script to reproduce the 48.5 AP on COCO zero-shot. See https://github.com/IDEA-Research/GroundingDINO#coco-zero-shot-evaluations for more details. Hope the file is helpful.

Thanks for your Python evaluation script.
However, the batch size hyper-parameter would have a bad effect on the evaluation results.

**Comment by zetaodu** [comment]
I still get 40.7 using the released evaluation code. Is the GroundingDINO_SwinT_OGC file misconfigured? Or is there a problem with the image processing? Size? Please help me! Thanks!

**Comment by zetaodu** [comment]
> 
I still get 40.7 using the released evaluation code. Is the GroundingDINO_SwinT_OGC file misconfigured? Or is there a problem with the image processing? Size? Please help me! Thanks!

**Comment by raja-kumar** [comment]
> > Hi @HuixinSun @serser @Dwrety . I provide a script to reproduce the 48.5 AP on COCO zero-shot. See https://github.com/IDEA-Research/GroundingDINO#coco-zero-shot-evaluations for more details. Hope the file is helpful.
> 
> Hi, @SlongLiu . I have tried your script and I find out that there is a category id mapping in the code, which make the whole category number increasing to 91. But coco only got 80 categories. What's that for?

Hi @CConory I am also having a hard time figuring out what that is. Did you figure this out?

**Comment by HuiZhang0812** [comment]
I ran the latest "test_ap_on_coco.py" code and still got all zeros. I don't know how to solve this problem.
<img width="614" alt="image" src="https://github.com/user-attachments/assets/d16e3b05-0acf-472f-8cb4-dc134b80f706">

**Comment by ztc20011103** [comment]
> I ran the latest "test_ap_on_coco.py" code and still got all zeros. I don't know how to solve this problem. <img alt="image" width="614" src="https://private-user-images.githubusercontent.com/44562591/382039147-d16e3b05-0acf-472f-8cb4-dc134b80f706.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM4ODAzMTUsIm5iZiI6MTczMzg4MDAxNSwicGF0aCI6Ii80NDU2MjU5MS8zODIwMzkxNDctZDE2ZTNiMDUtMGFjZi00NzJmLThjYjQtZGMxMzRiODBmNzA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjExVDAxMjAxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA1YmE1OGFhNWFhMzY4MjkxOTk1NjRiODZlZDI4M2Y3YWViYjliODlhN2Q2Nzg2OWRhZTgyZTA4MWY4MjQwZDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.CT9TyI122PYjdt9dQrYvdr2D6lTF9MvNL1jZ6hcO3GQ">

You can try this command. I've got 48.5ap on COCO Val2017.
CUDA_VISIBLE_DEVICES=0 \
python demo/test_ap_on_coco.py \
 -c groundingdino/config/GroundingDINO_SwinT_OGC.py \
 -p weights/groundingdino_swint_ogc.pth \
 --anno_path /path/to/annoataions/ie/instances_val2017.json \
 --image_dir /path/to/imagedir/ie/val2017
